webapp/advantage/decorators.py

```diff
@@ -144,18 +168,57 @@ def decorated_function(*args, **kwargs):
 
                     return flask.jsonify(message), 401
 
+            badgr_issuer = (
+                BADGR_ISSUER if not test_backend else QA_BADGR_ISSUER
+            )
+            certified_badge = (
+                CERTIFIED_BADGE if not test_backend else QA_CERTIFIED_BADGE,
+            )
+
             # init API instance
-            g.api = UAContractsAPI(
+            ua_api = UAContractsAPI(
                 session=session,
                 authentication_token=(user_token),
                 token_type=("Macaroon" if user_token else "Bearer"),
                 api_url=get_api_url(is_test_backend),
             )
 
-            if response == "html":
-                g.api.set_is_for_view(True)
+            badgr_api = BadgrAPI(
+                "https://api.eu.badgr.io"
+                if not test_backend
+                else "https://api.test.badgr.com",
+                os.getenv("BAGDR_USER"),
+                os.getenv("BADGR_PASSWORD")
+                if not test_backend
+                else os.getenv("BADGR_QA_PASSWORD"),
+                badgr_session,
+            )
 
-            return func(*args, **kwargs)
+            edx_api = EdxAPI(
+                "https://cube.ubuntu.com"
+                if not test_backend
+                else "https://qa.cube.ubuntu.com",
+                os.getenv("CUBE_EDX_CLIENT_ID")
+                if not test_backend
+                else os.getenv("CUBE_EDX_QA_CLIENT_ID"),
+                os.getenv("CUBE_EDX_CLIENT_SECRET")
+                if not test_backend
+                else os.getenv("CUBE_EDX_CLIENT_QA_SECRET"),
+                edx_session,
+            )
+
+            if response == "html":
+                ua_api.set_is_for_view(True)
+
+            return func(
+                badgr_issuer,
```

Comment:
Up to you, but I'd add `badgr_issuer`, `certified_badge`, `ua_api`, `edx_api` in `kwargs`, if you want to take this approach to provide these values to the view functions. That way you do not have to define these 5 arguments for every future view function definition.

Reply:
I think it's much clear this way, because you know exactly which are the parameters passed and the function will act as a guard if there are missing arguments.